h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,196 @@
+      final List<String> c = new ArrayList<>();
+      for (Bucket b : storage.list().iterateAll()) {
+        c.add(b.getName());
+      }
+      return c;
+    }
+  };
+
+  @Override
+  public List<String> calcTypeaheadMatches(String filter, int limit) {
+    final String input = GcsBlob.removePrefix(filter);
+    final String[] bk = input.split("/", 2);
+    if (bk.length == 1) {
+      return bucketCache.fetch(GcsBlob.KEY_PREFIX, bk[0], limit);
+    } else if (bk.length == 2) {
+      if (!keyCache.containsKey(bk[0])) {
+        StringCache sc = new StringCache() {
+          @Override
+          List<String> update() {
+            final List<String> cache = new ArrayList<>();
+            for (Blob b : storage.get(bk[0]).list().iterateAll()) {
+              cache.add(b.getName());
+            }
+            return cache;
+          }
+        };
+        keyCache.put(bk[0], sc);
+      }
+      return keyCache.get(bk[0]).fetch(GcsBlob.KEY_PREFIX + bk[0] + "/", bk[1], limit);
+    }
+    return new ArrayList<>(0);
+  }
+
+  @Override
+  public void importFiles(String path,
+                          String pattern,
+                          ArrayList<String> files,
+                          ArrayList<String> keys,
+                          ArrayList<String> fails,
+                          ArrayList<String> dels) {
+    // bk[0] is bucket name, bk[1] is file name - file name is optional.
+    final String bk[] = GcsBlob.removePrefix(path).split("/", 2);
+
+    if (bk.length < 2) {
+      parseBucket(bk[0], files, keys, fails);
+    } else {
+      try {
+        Blob blob = storage.get(bk[0], bk[1]);
+        final GcsBlob gcsBlob = GcsBlob.of(blob.getBlobId());
+        final Key k = GcsFileVec.make(path, blob.getSize());
+        keys.add(k.toString());
+        files.add(path);
+      } catch (Throwable t) {
+        fails.add(path);
+      }
+    }
+
+  }
+
+  private void parseBucket(String bucketId,
+                           ArrayList<String> files,
+                           ArrayList<String> keys,
+                           ArrayList<String> fails) {
+    final Bucket bucket = storage.get(bucketId);
+    for (Blob blob : bucket.list().iterateAll()) {
+        final GcsBlob gcsBlob = GcsBlob.of(blob.getBlobId());
+        Log.info("Importing: " + gcsBlob.toString());
+        try {
+          final Key k = GcsFileVec.make(gcsBlob.getCanonical(), blob.getSize());
+          keys.add(k.toString());
+          files.add(gcsBlob.getCanonical());
+        } catch (Throwable t) {
+          fails.add(gcsBlob.getCanonical());
+        }
+    }
+  }
+
+  @Override
+  public InputStream open(final String path) {
+    final GcsBlob gcsBlob = GcsBlob.of(path);
+    Log.info("Openning: " + gcsBlob.toString());
+    final Blob blob = storage.get(gcsBlob.getBlobId());
+    return new InputStream() {
+      @Override
+      public int read() throws IOException {
+        // very naive version with reading byte by byte
```

Comment:
gcsBlob.getBlobInfo().getSize(); provides information about length in bytes. No need to create ByteBuffer for each and every byte.